src/main/java/appeng/parts/p2p/PartP2PFEPower.java

```diff
@@ -73,180 +65,156 @@ public void onTunnelNetworkChange()
 	{
-		if( this.isOutput() )
+		if( capability == Capabilities.FORGE_ENERGY )
 		{
-			return 0;
+			return !this.isOutput();
 		}
+		return super.hasCapability( capability );
+	}
 
-		if( this.isActive() )
+	@Nullable
+	@Override
+	public <T> T getCapability( @Nonnull Capability<T> capability )
+	{
+		if( capability == Capabilities.FORGE_ENERGY )
 		{
-			int total = 0;
-
-			try
-			{
-				final int outputTunnels = this.getOutputs().size();
-
-				if( outputTunnels == 0 )
-				{
-					return 0;
-				}
-
-				final int amountPerOutput = maxReceive / outputTunnels;
-				int overflow = maxReceive % amountPerOutput;
-
-				for( PartP2PFEPower target : this.getOutputs() )
-				{
-					final IEnergyStorage output = target.getOutput();
-					final int toSend = amountPerOutput + overflow;
-					final int received = output.receiveEnergy( toSend, simulate );
-
-					overflow = toSend - received;
-					total += received;
-				}
-
-				this.queueTunnelDrain( PowerUnits.RF, total );
-			}
-			catch( GridAccessException ignored )
+			if( !this.isOutput() )
 			{
+				return (T) this.inputHandler;
 			}
-
-			return total;
+			return (T) NULL_ENERGY_STORAGE;
 		}
-
-		return 0;
+		return super.getCapability( capability );
 	}
 
-	@Override
-	public int extractEnergy( int maxExtract, boolean simulate )
+	private class InputEnergyStorage implements IEnergyStorage
 	{
-		return 0;
-	}
+		@Override
+		public int extractEnergy( int maxExtract, boolean simulate )
+		{
+			return 0;
+		}
 
-	private IEnergyStorage getOutput()
-	{
-		if( this.isOutput() )
+		@Override
+		public int receiveEnergy( int maxReceive, boolean simulate )
 		{
-			if( !this.cachedTarget )
+			if( PartP2PFEPower.this.isActive() )
 			{
-				final TileEntity self = this.getTile();
-				final TileEntity te = self.getWorld().getTileEntity( new BlockPos( self.getPos().getX() + this.getSide().xOffset, self.getPos()
-						.getY() + this.getSide().yOffset, self.getPos().getZ() + this.getSide().zOffset ) );
+				int total = 0;
 
-				if( te != null && te.hasCapability( Capabilities.FORGE_ENERGY, this.getSide().getOpposite().getFacing() ) )
+				try
 				{
-					this.outputTarget = te.getCapability( Capabilities.FORGE_ENERGY, this.getSide().getOpposite().getFacing() );
+					final int outputTunnels = PartP2PFEPower.this.getOutputs().size();
+
+					if( outputTunnels == 0 )
+					{
+						return 0;
+					}
+
+					final int amountPerOutput = maxReceive / outputTunnels;
+					int overflow = maxReceive % amountPerOutput;
+
+					for( PartP2PFEPower target : PartP2PFEPower.this.getOutputs() )
+					{
+						final IEnergyStorage output = target.getOutput();
+						final int toSend = amountPerOutput + overflow;
+						final int received = output.receiveEnergy( toSend, simulate );
+
+						overflow = toSend - received;
+						total += received;
+					}
+
+					PartP2PFEPower.this.queueTunnelDrain( PowerUnits.RF, total );
 				}
-				else
+				catch( GridAccessException ignored )
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Either log or rethrow this exception. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1166)